src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,153 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    internal class RemoteBrowserFileStream : BrowserFileStream
+    {
+        private readonly IJSRuntime _jsRuntime;
+        private readonly ElementReference _inputFileElement;
+        private readonly int _maxChunkSize;
+        private readonly PipeReader _pipeReader;
+        private readonly CancellationTokenSource _fillBufferCts;
+
+        private bool _isReadingCompleted;
+        private bool _isDisposed;
+
+        public RemoteBrowserFileStream(IJSRuntime jsRuntime, ElementReference inputFileElement, int maxChunkSize, int maxBufferSize, BrowserFile file)
+            : base(file)
+        {
+            _jsRuntime = jsRuntime;
+            _inputFileElement = inputFileElement;
+            _maxChunkSize = maxChunkSize;
+
+            var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize));
+            _pipeReader = pipe.Reader;
+            _fillBufferCts = new CancellationTokenSource();
+
+            _ = FillBuffer(pipe.Writer, _fillBufferCts.Token);
+        }
+
+        private async Task FillBuffer(PipeWriter writer, CancellationToken cancellationToken)
+        {
+            long offset = 0;
+
+            while (offset < File.Size)
+            {
+                var pipeBuffer = writer.GetMemory(_maxChunkSize);
+                var chunkSize = (int)Math.Min(_maxChunkSize, File.Size - offset);
+
+                try
+                {
+                    var base64 = await _jsRuntime.InvokeAsync<string>(
+                        InputFileInterop.ReadFileData,
+                        cancellationToken,
+                        _inputFileElement,
+                        File.Id,
+                        offset,
+                        chunkSize);
+
+                    if (!Convert.TryFromBase64String(base64, pipeBuffer.Span, out var bytesWritten))
+                    {
+                        throw new FormatException("A chunk with an invalid format was received.");
+                    }
+
+                    if (bytesWritten != chunkSize)
+                    {
+                        throw new InvalidOperationException(
+                            $"A chunk with size {bytesWritten} bytes was received, but {chunkSize} bytes were expected.");
+                    }
+
+                    writer.Advance(chunkSize);
+                    offset += chunkSize;
+                }
+                catch (Exception e)
+                {
+                    await writer.CompleteAsync(e);
+                    throw;
+                }
+
+                var result = await writer.FlushAsync();
```

Comment:
This call to FlushAsync should probably take the cancellationToken. CompleteAsync() should also be called in a finally block in case FlushAsync throws.